File: backend/apps/trading/models/brokers.py
"""
Modèles liés aux brokers et connexions API.
"""
from django.db import models
from django.contrib.auth.models import User
from .base import TimeStampedModel


class Broker(TimeStampedModel):
    """Broker (courtier) supporté."""
    
    class BrokerType(models.TextChoices):
        SAXO = 'SAXO', 'Saxo Bank'
        BINANCE = 'BINANCE', 'Binance'
        INTERACTIVE_BROKERS = 'IB', 'Interactive Brokers'
        OTHER = 'OTHER', 'Autre'
    
    name = models.CharField(max_length=100)
    broker_type = models.CharField(
        max_length=20, choices=BrokerType.choices
    )
    
    # Configuration API
    api_base_url = models.URLField(blank=True)
    api_version = models.CharField(max_length=20, blank=True)
    
    is_active = models.BooleanField(default=True)
    supports_stocks = models.BooleanField(default=True)
    supports_crypto = models.BooleanField(default=False)
    supports_forex = models.BooleanField(default=False)
    supports_futures = models.BooleanField(default=False)
    
    class Meta:
        ordering = ['name']
        verbose_name = 'Broker'
        verbose_name_plural = 'Brokers'
    
    def __str__(self):
        return self.name


class BrokerAccount(TimeStampedModel):
    """
    Compte utilisateur chez un broker.
    Architecture inspirée de BrokerCredentials de la v3.
    """
    user = models.ForeignKey(
        User, on_delete=models.CASCADE, related_name='broker_accounts'
    )
    broker = models.ForeignKey(
        Broker, on_delete=models.CASCADE, related_name='accounts', null=True, blank=True
    )
    
    # Nom pour identifier cette configuration (comme dans v3)
    name = models.CharField(
        max_length=100,
        default='',
        blank=True,
        help_text="Nom pour identifier cette configuration (ex: 'Mon compte Saxo principal')"
    )
    
    # Type de broker (redondant avec broker.broker_type mais utile pour compatibilité v3)
    broker_type = models.CharField(
        max_length=20,
        choices=Broker.BrokerType.choices,
        default=Broker.BrokerType.OTHER,
        help_text="Type de broker (Saxo, Binance, etc.)"
    )
    
    # Environnement de trading
    ENVIRONMENT_CHOICES = [
        ('live', 'Live Trading'),
        ('simulation', 'Simulation/Demo'),
    ]
    environment = models.CharField(
        max_length=20, 
        choices=ENVIRONMENT_CHOICES, 
        default='live',  # ✅ MIGRATION: Défaut changé de 'simulation' vers 'live'
        help_text="Environnement de trading (Live ou Simulation)"
    )
    
    # Account ID (optionnel, pour identifier le compte chez le broker)
    account_id = models.CharField(max_length=100, blank=True, help_text="ID du compte chez le broker")
    
    # ============================================
    # CREDENTIALS SAXO BANK
    # ============================================
    saxo_client_id = models.CharField(max_length=255, blank=True, null=True)
    saxo_client_secret = models.CharField(max_length=255, blank=True, null=True)
    saxo_redirect_uri = models.URLField(blank=True, null=True)
    saxo_environment = models.CharField(
        max_length=20,
        choices=[('live', 'Live'), ('simulation', 'Simulation')],
        default='live',  # ✅ MIGRATION: Défaut changé de 'simulation' vers 'live'
        verbose_name="Environnement Saxo"
    )
    saxo_access_token = models.TextField(blank=True, null=True)
    saxo_refresh_token = models.TextField(blank=True, null=True)
    saxo_token_expires_at = models.DateTimeField(blank=True, null=True)
    
    # ============================================
    # CREDENTIALS BINANCE
    # ============================================
    binance_api_key = models.CharField(max_length=255, blank=True, null=True)
    binance_api_secret = models.CharField(max_length=255, blank=True, null=True)
    binance_testnet = models.BooleanField(default=False)
    
    # ============================================
    # CREDENTIALS GÉNÉRIQUES (pour autres brokers)
    # ============================================
    # API Keys génériques (pour compatibilité avec l'ancien modèle)
    api_key = models.CharField(max_length=255, blank=True)
    api_secret = models.CharField(max_length=255, blank=True)
    client_id = models.CharField(max_length=255, blank=True)
    client_secret = models.CharField(max_length=255, blank=True)
    access_token = models.TextField(blank=True)
    refresh_token = models.TextField(blank=True)
    token_expires_at = models.DateTimeField(null=True, blank=True)
    
    # Extra credentials (JSON pour données supplémentaires)
    extra_credentials = models.JSONField(default=dict, blank=True)
    
    # ============================================
    # CONFIGURATION AUTO-REFRESH (Saxo)
    # ============================================
    auto_refresh_enabled = models.BooleanField(
        default=True, 
        verbose_name="Auto-refresh activé",
        help_text="Activer le rafraîchissement automatique des tokens"
    )
    auto_refresh_frequency = models.IntegerField(
        default=45,
        verbose_name="Fréquence auto-refresh (minutes)",
        help_text="Fréquence en minutes (15-55 min)"
    )
    
    # ============================================
    # BALANCE ET STATUT
    # ============================================
    balance = models.DecimalField(
        max_digits=20, decimal_places=2, null=True, blank=True
    )
    currency = models.CharField(max_length=10, default='EUR')
    balance_updated_at = models.DateTimeField(null=True, blank=True)
    last_sync = models.DateTimeField(null=True, blank=True, help_text="Dernière synchronisation")
    
    is_active = models.BooleanField(default=True)
    is_demo = models.BooleanField(default=False)
    is_sandbox = models.BooleanField(
        default=True, 
        help_text="Use sandbox/simulation environment"
    )
    
    class Meta:
        unique_together = ['user', 'broker_type', 'name']
        ordering = ['broker_type', 'name']
        verbose_name = 'Broker Account'
        verbose_name_plural = 'Broker Accounts'
    
    def __str__(self):
        return f"{self.user.username} - {self.get_broker_type_display()} - {self.name}"
    
    def get_broker_type(self) -> str:
        """
        Récupère le broker_type de manière sécurisée.
        Utilise le champ broker_type directement, ou fallback sur broker.broker_type.
        
        Returns:
            Type de broker (SAXO, BINANCE, etc.)
        """
        if self.broker_type:
            return self.broker_type
        elif self.broker and self.broker.broker_type:
            # Mettre à jour le champ pour éviter de refaire cette vérification
            self.broker_type = self.broker.broker_type
            self.save(update_fields=['broker_type'])
            return self.broker_type
        else:
            return 'OTHER'
    
    def get_credentials_dict(self) -> dict:
        """
        Retourner les credentials sous forme de dictionnaire.
        Compatible avec l'architecture v3.
        """
        broker_type = self.get_broker_type()
        
        if broker_type == 'SAXO':
            return {
                'client_id': self.saxo_client_id or self.client_id,
                'client_secret': self.saxo_client_secret or self.client_secret,
                'redirect_uri': self.saxo_redirect_uri,
                'access_token': self.saxo_access_token or self.access_token,
                'refresh_token': self.saxo_refresh_token or self.refresh_token,
                'token_expires_at': self.saxo_token_expires_at or self.token_expires_at,
                # ✅ MIGRATION: Pour Saxo, privilégier saxo_environment, sinon environment, sinon défaut 'live'
                'environment': (self.saxo_environment if hasattr(self, 'saxo_environment') and self.saxo_environment else None) or self.environment or 'live',
            }
        elif broker_type == 'BINANCE':
            return {
                'api_key': self.binance_api_key or self.api_key,
                'api_secret': self.binance_api_secret or self.api_secret,
                'testnet': self.binance_testnet if hasattr(self, 'binance_testnet') else self.is_sandbox,
                'environment': self.environment,
            }
        else:
            # Pour les autres brokers, utiliser les champs génériques
            return {
                'api_key': self.api_key,
                'api_secret': self.api_secret,
                'client_id': self.client_id,
                'client_secret': self.client_secret,
                'access_token': self.access_token,
                'refresh_token': self.refresh_token,
                'token_expires_at': self.token_expires_at,
                'environment': self.environment or 'simulation',
            }


class BrokerSyncLog(TimeStampedModel):
    """Log de synchronisation avec un broker."""
    
    class SyncType(models.TextChoices):
        POSITIONS = 'POSITIONS', 'Positions'
        TRADES = 'TRADES', 'Trades'
        ORDERS = 'ORDERS', 'Ordres'
        BALANCE = 'BALANCE', 'Balance'
        ASSETS = 'ASSETS', 'Assets'
        PRICES = 'PRICES', 'Prix'
        CONNECTION_TEST = 'CONNECTION_TEST', 'Test connexion'
        ORDER_PLACED = 'ORDER_PLACED', 'Ordre placé'
        ORDER_CANCELLED = 'ORDER_CANCELLED', 'Ordre annulé'
    
    class SyncStatus(models.TextChoices):
        SUCCESS = 'SUCCESS', 'Succès'
        PARTIAL = 'PARTIAL', 'Partiel'
        FAILED = 'FAILED', 'Échec'
        ERROR = 'ERROR', 'Erreur'
        WARNING = 'WARNING', 'Avertissement'
    
    broker_account = models.ForeignKey(
        BrokerAccount, on_delete=models.CASCADE, related_name='sync_logs'
    )
    sync_type = models.CharField(max_length=50)  # Allow custom sync types
    status = models.CharField(max_length=20, choices=SyncStatus.choices)
    
    records_synced = models.IntegerField(default=0)
    items_synced = models.IntegerField(default=0)  # Legacy field
    error_message = models.TextField(blank=True, default='')
    details = models.JSONField(default=dict, blank=True)
    
    started_at = models.DateTimeField(auto_now_add=True)
    completed_at = models.DateTimeField(null=True, blank=True)
    
    class Meta:
        ordering = ['-created_at']
        verbose_name = 'Broker Sync Log'
        verbose_name_plural = 'Broker Sync Logs'
    
    def __str__(self):
        return f"{self.broker_account} - {self.sync_type} ({self.status})"

File: frontend/src/services/brokers.ts
/**
 * Service pour les Brokers
 */
import apiClient from './api/client'
import type { ApiResponse, Broker, BrokerAccount, BrokerSyncLog, ApiError } from '@types'

export interface BrokerAccountCreateData {
  // Identifiants
  name: string
  broker?: number | null
  broker_type: 'SAXO' | 'BINANCE' | 'IB' | 'OTHER'
  account_id?: string
  environment?: 'live' | 'simulation'
  
  // Credentials Saxo
  saxo_client_id?: string
  saxo_client_secret?: string
  saxo_redirect_uri?: string
  saxo_environment?: 'live' | 'simulation'
  saxo_access_token?: string
  saxo_refresh_token?: string
  saxo_token_expires_at?: string | null
  
  // Credentials Binance
  binance_api_key?: string
  binance_api_secret?: string
  binance_testnet?: boolean
  
  // Credentials génériques (pour compatibilité)
  api_key?: string
  api_secret?: string
  client_id?: string
  client_secret?: string
  access_token?: string
  refresh_token?: string
  token_expires_at?: string | null
  extra_credentials?: Record<string, any>
  
  // Auto-refresh
  auto_refresh_enabled?: boolean
  auto_refresh_frequency?: number
  
  // Statut
  is_active?: boolean
  is_demo?: boolean
  is_sandbox?: boolean
}

export interface TestConnectionResponse {
  success: boolean
  message: string
  error?: string
}

export interface SyncRequest {
  sync_type: 'ASSETS' | 'PRICES' | 'POSITIONS' | 'TRADES'
  force?: boolean
}

export const brokerService = {
  /**
   * Récupérer tous les brokers
   */
  async getAll(): Promise<ApiResponse<Broker>> {
    const response = await apiClient.get<ApiResponse<Broker>>('/brokers/')
    return response.data
  },

  /**
   * Récupérer un broker par ID
   */
  async getById(id: number): Promise<Broker> {
    const response = await apiClient.get<Broker>(`/brokers/${id}/`)
    return response.data
  },

  /**
   * Récupérer tous les comptes broker
   */
  async getAccounts(): Promise<ApiResponse<BrokerAccount>> {
    const response = await apiClient.get<ApiResponse<BrokerAccount>>('/broker-accounts/')
    return response.data
  },

  /**
   * Récupérer un compte broker par ID
   */
  async getAccountById(id: number): Promise<BrokerAccount> {
    const response = await apiClient.get<BrokerAccount>(`/broker-accounts/${id}/`)
    return response.data
  },

  /**
   * Créer un compte broker
   */
  async createAccount(data: BrokerAccountCreateData): Promise<BrokerAccount> {
    const response = await apiClient.post<BrokerAccount>('/broker-accounts/', data)
    return response.data
  },

  /**
   * Mettre à jour un compte broker
   */
  async updateAccount(id: number, data: Partial<BrokerAccountCreateData>): Promise<BrokerAccount> {
    const response = await apiClient.patch<BrokerAccount>(`/broker-accounts/${id}/`, data)
    return response.data
  },

  /**
   * Supprimer un compte broker
   */
  async deleteAccount(id: number): Promise<void> {
    await apiClient.delete(`/broker-accounts/${id}/`)
  },

  /**
   * Tester la connexion à un broker
   */
  async testConnection(accountId: number): Promise<TestConnectionResponse> {
    const response = await apiClient.post<TestConnectionResponse>(
      `/broker-accounts/${accountId}/test-connection/`
    )
    return response.data
  },

  /**
   * Récupérer le solde EUR d'un compte broker
   */
  async getBalanceEur(accountId: number): Promise<{
    success: boolean
    balance_eur: number
    currency: string
    all_balances: Record<string, number>
    timestamp?: string
    error?: string
  }> {
    const response = await apiClient.get<{
      success: boolean
      balance_eur: number
      currency: string
      all_balances: Record<string, number>
      timestamp?: string
      error?: string
    }>(`/broker-accounts/${accountId}/balance-eur/`)
    return response.data
  },

  /**
   * Rafraîchir le solde d'un compte broker
   */
  async refreshBalance(accountId: number): Promise<{
    success: boolean
    balance_eur: number
    currency: string
    all_balances: Record<string, number>
    account: BrokerAccount
    error?: string
  }> {
    const response = await apiClient.post<{
      success: boolean
      balance_eur: number
      currency: string
      all_balances: Record<string, number>
      account: BrokerAccount
      error?: string
    }>(`/broker-accounts/${accountId}/refresh-balance/`)
    return response.data
  },

  /**
   * Récupérer les credentials (masqués) d'un compte broker
   */
  async getCredentials(accountId: number): Promise<{
    broker_type: string
    credentials_dict: Record<string, any>
    raw_fields: Record<string, any>
    has_api_key: boolean
    has_api_secret: boolean
    testnet: boolean
    environment: string
  }> {
    const response = await apiClient.get<{
      broker_type: string
      credentials_dict: Record<string, any>
      raw_fields: Record<string, any>
      has_api_key: boolean
      has_api_secret: boolean
      testnet: boolean
      environment: string
    }>(`/broker-accounts/${accountId}/credentials/`)
    return response.data
  },

  /**
   * Obtenir l'URL d'authentification Saxo (OAuth2)
   */
  async getSaxoAuthUrl(accountId: number): Promise<{ auth_url: string; state: string }> {
    const response = await apiClient.get<{ success: boolean; auth_url: string; state: string }>(
      `/broker-accounts/${accountId}/saxo-auth-url/`
    )
    return response.data
  },

  /**
   * Échanger le code d'autorisation Saxo contre des tokens
   */
  async exchangeSaxoAuthCode(accountId: number, code: string, state?: string): Promise<BrokerAccount> {
    const response = await apiClient.post<BrokerAccount>(`/broker-accounts/${accountId}/saxo-exchange-code/`, {
      code,
      state,
    })
    return response.data
  },

  /**
   * Rafraîchir le token Saxo
   */
  async refreshSaxoToken(accountId: number): Promise<BrokerAccount> {
    const response = await apiClient.post<BrokerAccount>(`/broker-accounts/${accountId}/saxo-refresh-token/`)
    return response.data
  },

  /**
   * Supprimer les tokens Saxo
   */
  async deleteSaxoTokens(accountId: number): Promise<BrokerAccount> {
    const response = await apiClient.post<BrokerAccount>(`/broker-accounts/${accountId}/saxo-delete-tokens/`)
    return response.data
  },

  /**
   * Récupérer les assets depuis Saxo
   */
  async getSaxoAssets(
    accountId: number,
    options?: {
      asset_type?: string
      keywords?: string
      limit?: number
    }
  ): Promise<{
    success: boolean
    count: number
    assets: Array<{
      symbol: string
      name: string
      asset_type: string
      exchange: string
      currency: string
      is_tradable: boolean
      broker_id: string | null
    }>
    error?: string
  }> {
    const params = new URLSearchParams()
    if (options?.asset_type) params.append('asset_type', options.asset_type)
    if (options?.keywords) params.append('keywords', options.keywords)
    if (options?.limit) params.append('limit', options.limit.toString())

    try {
      const response = await apiClient.get<{
        success: boolean
        count: number
        assets: Array<{
          symbol: string
          name: string
          asset_type: string
          exchange: string
          currency: string
          is_tradable: boolean
          broker_id: string | null
        }>
      }>(`/broker-accounts/${accountId}/saxo-assets/?${params.toString()}`)
      return response.data
    } catch (error: any) {
      return {
        success: false,
        count: 0,
        assets: [],
        error: error.response?.data?.error || error.message || 'Failed to fetch Saxo assets',
      }
    }
  },

  /**
   * Récupérer les positions depuis Saxo
   */
  async getSaxoPositions(accountId: number): Promise<{
    success: boolean
    count: number
    positions: Array<{
      symbol: string
      quantity: number
      entry_price: number
      current_price: number | null
      unrealized_pnl: number | null
      currency: string
      side: string
      broker_id: string | null
    }>
    error?: string
  }> {
    try {
      const response = await apiClient.get<{
        success: boolean
        count: number
        positions: Array<{
          symbol: string
          quantity: number
          entry_price: number
          current_price: number | null
          unrealized_pnl: number | null
          currency: string
          side: string
          broker_id: string | null
        }>
      }>(`/broker-accounts/${accountId}/saxo-positions/`)
      return response.data
    } catch (error: any) {
      return {
        success: false,
        count: 0,
        positions: [],
        error: error.response?.data?.error || error.message || 'Failed to fetch Saxo positions',
      }
    }
  },

  /**
   * Synchroniser les données depuis un broker
   */
  async sync(accountId: number, syncRequest: SyncRequest): Promise<BrokerSyncLog> {
    const response = await apiClient.post<BrokerSyncLog>(
      `/broker-accounts/${accountId}/sync/`,
      syncRequest
    )
    return response.data
  },

  /**
   * Récupérer les logs de synchronisation
   */
  async getSyncLogs(accountId?: number): Promise<ApiResponse<BrokerSyncLog>> {
    const params = accountId ? { broker_account: accountId } : {}
    const response = await apiClient.get<ApiResponse<BrokerSyncLog>>('/broker-sync-logs/', {
      params,
    })
    return response.data
  },

  /**
   * Récupérer le dernier log de synchronisation
   */
  async getLastSyncLog(accountId: number, syncType?: string): Promise<BrokerSyncLog | null> {
    const params: Record<string, any> = { broker_account: accountId }
    if (syncType) {
      params.sync_type = syncType
    }
    
    const response = await apiClient.get<ApiResponse<BrokerSyncLog>>('/broker-sync-logs/', {
      params: { ...params, page_size: 1, ordering: '-started_at' },
    })
    
    return response.data.results[0] || null
  },
}

export default brokerService

